### PR TITLE
Adding option to use a fork of showyourwork

### PR DESCRIPTION
**Pull request by dfm**
I think that this should be backwards compatible, but it permits the following syntax:

```yaml
version:
  fork: https://github.com/dfm/showyourwork.git
  spec: some-branch-tag-or-commit-hash
```

Omitting the `spec` argument will use the default branch, and omitting the `fork` argument will use the base `showyourwork/showyourwork` repo.

Ref: #145 for some context

### CODE REVIEW

**Comment by dfm** [comment]
I can confirm that this seems to work as expected for my project, although I don't know how to include an integration test.

File: docs/config.rst
Comment (rodluger) [inline docs/config.rst]: typo here, should be `version`

File: docs/config.rst
Comment (rodluger) [inline docs/config.rst]: typo here, should be `version`

**Comment by rodluger** [comment]
Cool, thanks! Just wondering: why not keep it simple and let the user provide something like

```
version: https://github.com/user/repo@spec
```

instead of the `fork:` and `spec:` logic?

**Comment by dfm** [comment]
I think the regex logic in there already is pretty gnarly so this seemed easier to maintain in the long term, but I'd be happy to use that mode if you prefer!

**Comment by rodluger** [comment]
I agree your approach is more elegant. It also opens the door to allowing arbitrary branch names in the showyourwork/showyourwork repo (currently I only allow `main` and `dev` since other branches could in principle look like relative path names on disk). This would let us do something like

```
version:
    branch: foobar
```

So it might be useful to take a step back now and figure out a better syntax for all possible version specifications:

- PyPI version
    ```
    version:
        pip: 0.3.0
    ```
- GitHub SHA
    ```
    version:
        sha: abcde
    ```
- GitHub branch
    ```
    version:
        branch: main
    ```
- GitHub SHA on fork
    ```
    version:
        fork: dfm/showyourwork # or github.com/dfm/showyourwork
        sha: abcde # or branch: main
    ```
- Local
    ```
    version:
        path: /path/to/local/folder
    ```

And, of course, all backwards-compatible version specs. So we'd allow either a string (current behavior), `pip`, or one of (`path`, `sha`, or `branch`). For `sha` and `branch`, we further allow an optional `fork` key that defaults to `showyourwork/showyourwork`.

What do you think?

**Comment by dfm** [comment]
I think this is great - Happy to implement!

**Comment by rodluger** [comment]
Yes please! As for testing, the current approach is fine. We can integrate it into the full testing workflow tomorrow when we chat.

**Comment by rodluger** [comment]
(BTW, is the url structure `<baseurl>/user/repo.git@spec#egg=egg` universal? Is that the case for bitbucket, gitlab, etc.?)

**Comment by dfm** [comment]
All of the above work now plus also `tag` which should have the same syntax. A different option would be to just use `ref` instead of `branch`, `tag`, and `sha` since they all have the same behavior and I think `ref` is the usual name...

**Comment by dfm** [comment]
I think that this syntax would work for other `git`-based platforms, yes!

**Comment by dfm** [comment]
I did switch to `ref` to cover `branch`, `tag` and `sha`. If you'd like to include those as aliases, we can add them back in.

**Comment by rodluger** [comment]
Yeah, `ref` is good -- thanks!

**Comment by rodluger** [comment]
BTW, I don't have Actions set up to run tests on PRs because the testing workflow is expensive and spins up other actions workflows in all the test repositories (`showyourwork/test-***`). Would be nice to allow the user to request a workflow run by, e.g., posting a PR comment and having a bot spin up the test after admin approval -- or something like that. What do you think @dfm?

**Comment by dfm** [comment]
That would be very cool!! 🤖

**Comment by rodluger** [comment]
@dfm can you rebase and commit to see if it triggers a workflow run?

**Comment by rodluger** [comment]
Nice. The new strategy for running PR workflows is working! (In brief: I use a `pull_request_target` trigger, since the workflow needs to be able to write to the test repos in the `showyourwork` org, but I execute the workflow within an isolated [environment](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) that requires admin approval to run).

**Comment by dfm** [comment]
Sweet!!

**Review by rodluger — APPROVED** [review]

**Comment by rodluger** [comment]
Thanks for implementing this!

**Pull request merged**
